src/riscv/lib/src/jit/builder/sequence.rs

```diff
@@ -205,6 +237,52 @@ impl<'jit, MC: MemoryConfig> SequenceBuilder<'jit, MC> {
         self.builder.ins().return_(&[steps]);
     }
 
+    /// IR for calling the interpreted mode fallback function when unable to continue in JIT mode.
+    /// This exit block writes the current PC back to the machine core state before calling the
+    /// interpreter, and returns the total number of steps executed in JIT and interpreted mode.
+    ///
+    /// TODO RV-842: Introduce counter update here for number of calls to fallback to interpreter.
+    fn fill_fallback_exit_block(&mut self) {
+        self.builder.switch_to_block(self.fallback_exit_block);
+        self.builder.set_cold_block(self.fallback_exit_block);
+
+        // SAFETY: The `current_pc` parameter is always a valid `Address` value.
+        let current_pc = unsafe {
+            let raw_value = self.builder.block_params(self.fallback_exit_block)[0];
+            Value::<Address>::from_raw(raw_value)
+        };
+
+        write_pc(self, current_pc);
+
+        let steps_remaining = self.builder.block_params(self.fallback_exit_block)[1];
+
+        let init_max_steps = self.max_steps_param.to_value();
+        let jit_steps_taken = self.builder.ins().isub(init_max_steps, steps_remaining);
+
+        // SAFETY: The `jit_steps_taken` should always be a valid `usize` value.
```

Comment:
nit - `should` is not strong enough to justify an `unsafe` block. We must be able to guarantee that requirements are upheld